File: boggle.cc
#include <algorithm>
#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#define NUM_MOVES 8
using namespace std;

const int moves[NUM_MOVES][2] =
{
    {-1, -1}, {-1, 0}, {-1, 1},
    {0, -1}, {0, 1},
    {1, -1}, {1, 0}, {1, 1}
};

inline bool cmp(string a, string b) {
    return a.size() < b.size();
}

bool match(vector<string> board, string word);
bool search(vector<string> board, vector< vector<bool> > visited,
        int i, int j, string word);

int main(int argc, char *argv[]) {
    string filename = "wordlist";
    if (argc >= 2)
        filename = argv[1];

    ifstream dict(filename.c_str());

    vector<string> words;

    int n, m;
    cin >> n >> m;

    vector<string> board(n);
    for (int i = 0; i < n; i++)
        cin >> board[i];

    string word;
    for (dict >> word; !dict.eof(); dict >> word)
        if (word.size() >= m && match(board, word))
            words.push_back(word);

    sort(words.rbegin(), words.rend(), cmp);

    for (vector<string>::iterator it = words.begin(); it != words.end(); ++it)
        cout << *it << endl;

    dict.close();
}

bool match(vector<string> board, string word) {
    int dim = board.size();
    vector< vector<bool> > visited(dim);
    for (int i = 0; i < dim; i++)
        visited[i].resize(dim);

    for (int i = 0; i < dim; i++)
        for (int j = 0; j < dim; j++)
            if (search(board, visited, i, j, word))
                return true;
    return false;
}

bool search(vector<string> board, vector< vector<bool> > visited,
        int i, int j, string word) {
    int dim = board.size();
    if (word.size() == 0)
        return true;
    else if (i < 0 || i >= dim || j < 0 || j >= dim)
        return false;
    else if (visited[i][j])
        return false;
    else if (board[i][j] != word[0])
        return false;
    else if (word[0] == 'q' && word[1] != 'u')
        return false;
    visited[i][j] = true;
    for (int m = 0; m < NUM_MOVES; m++)
        if (search(board, visited, i + moves[m][0], j + moves[m][1],
                word[0] == 'q' ? word.substr(2) : word.substr(1)))
            return true;
    return false;
}

File: boggle.pl
#!/usr/bin/perl -w

use strict;
use Data::Dumper;

use constant moves => (
    [-1, -1], [-1, 0], [-1, 1],
    [0, -1], [0, 1],
    [1, -1], [1, 0], [1, 1],
);

my $filename = "twl06.txt";
if (@ARGV >= 1) {
    $filename = shift @ARGV;
}

open my $dict, '<', $filename;

sub blank {
    my $rows = shift;
    my $cols = shift;
    my @arr = ();
    my @row = ();

    push @row, '' for (1..$cols);
    push @arr, [@row] for (1..$rows);

    return \@arr;
}

sub match {
    my $ref = shift;                    # ref to boggle board
    my $word = shift;                   # word to match

    my $rows = $#$ref;
    return '' if $rows < 1;
    my $cols = $#{$ref->[0]};

    for my $i (0..$rows) {
        for my $j (0..$cols) {
            return 1 if search($ref, blank($rows + 1, $cols + 1),
                [$rows, $cols], [$i, $j], $word);
        }
    }

    return '';                          # undef is probably implicitly, anyway
}

sub search {  ## TODO FIX Qu behavior
    my $ref = shift;                    # ref to boggle board
    my $vref = shift;                   # ref to visited
    my $dref = shift;                   # ref to anon array of dimensions
    my $iref = shift;                   # ref to anon array of indices
    my $word = shift;                   # word to search form

    return 1 if 0 == length $word;
    return '' if $iref->[0] < 0 || $iref->[0] > $dref->[0] ||
                $iref->[1] < 0 || $iref->[1] > $dref->[1];
    return '' if $vref->[$iref->[0]][$iref->[1]];
    return '' if $ref->[$iref->[0]][$iref->[1]] ne substr $word, 0, 1;

    $vref->[$iref->[0]][$iref->[1]] = 1;

    for my $move (moves) {
        return 1 if search($ref, $vref, $dref,
            [$iref->[0] + $move->[0], $iref->[1] + $move->[1]],
            (substr $word, 1));
    }

    $vref->[$iref->[0]][$iref->[1]] = '';

    return '';
}

my ($n, $m) = split ' ', <>;
my @grid = ();
for (1..$n) {
    my $line = <>;
    chomp $line;
    push @grid, [split //, $line];
}

my @arr = ();

while (<$dict>) {
    chomp;
    push @arr, $_ if length $_ >= $m && match(\@grid, $_);
}

for my $word (sort { length $b <=> length $a } @arr) {
    print $word, "\n";
}
